### PR TITLE
DCS-1516 Adding 'expected' field to new arrival

### PR DESCRIPTION
**Pull request by andrewrlee**

### CODE REVIEW

File: server/routes/__testutils/appSetup.ts [hunk 1]
```diff
@@ -23,7 +25,13 @@ export const user = {
   authSource: 'NOMIS',
 }
 
-export const signedCookiesProvider = jest.fn()
+const signedCookiesProvider = jest.fn()
+
+export const stubCookie = <T>(state: StateOperations<T>, value: T) =>
+  stubRequestCookie(signedCookiesProvider, state, value)
+
+export const stubCookies = (cookies: ExpectedCookie<unknown>[]) => stubRequestCookies(signedCookiesProvider, cookies)
```
Comment (andrewrlee) [inline server/routes/__testutils/appSetup.ts:33]: couple of new util methods which mean we don't have to refer to cookie name everywhere

File: server/routes/__testutils/requestTestUtils.ts [hunk 2]
```diff
@@ -69,13 +70,41 @@ export const mockResponse = ({ locals = { context: {}, user: {} } }: ResponsePar
...
+  signedCookiesProvider: jest.Mock<any, any>,
+  cookies: ExpectedCookie<unknown>[]
+) => {
+  signedCookiesProvider.mockReturnValue(
+    cookies.reduce((acc, [stateOperations, value]) => {
+      acc[stateOperations.cookieName] = stateOperations.write(value)
+      return acc
+    }, {})
+  )
+}
```
Comment (andrewrlee) [inline server/routes/__testutils/requestTestUtils.ts:110]: impls of helper methods

File: server/routes/bookedtoday/arrivals/autoMatchingRecords/multipleMatchingRecordsFoundController.test.ts [hunk 3]
```diff
@@ -102,10 +104,18 @@ describe('possible records found', () => {
         .send({ prisonNumber: arrival.prisonNumber })
         .expect(302)
         .expect('Location', '/prisoners/12345-67890/sex')
-        .expect(() => {
+        .expect(req => {
           expect(flashProvider).not.toHaveBeenCalled()
           expect(expectedArrivalsService.getPrisonerDetails).toHaveBeenCalledWith('A1234BC')
-          expect(signedCookiesProvider).toHaveBeenCalledTimes(1)
```
Comment (andrewrlee) [inline server/routes/bookedtoday/arrivals/autoMatchingRecords/multipleMatchingRecordsFoundController.test.ts:108]: a lot of these tests were'nt actually testing what state was being persisted

File: server/routes/bookedtoday/arrivals/autoMatchingRecords/noMatchingRecordsFoundController.test.ts [hunk 4]
```diff
@@ -1,6 +1,6 @@
 import type { Express } from 'express'
 import request from 'supertest'
-import cheerio from 'cheerio'
+import * as cheerio from 'cheerio'
```
Comment (andrewrlee) [inline server/routes/bookedtoday/arrivals/autoMatchingRecords/noMatchingRecordsFoundController.test.ts:3]: fixing lots of deprecated warnings

File: server/routes/bookedtoday/arrivals/reviewPerDetailsController.ts [hunk 5]
```diff
@@ -16,6 +16,7 @@ export default class ReviewPerDetailsController {
       sex: arrival.gender,
       pncNumber: arrival.pncNumber,
       prisonNumber: arrival.prisonNumber,
+      expected: true,
```
Comment (andrewrlee) [inline server/routes/bookedtoday/arrivals/reviewPerDetailsController.ts:19]: setting new flag to true when its an expected arrival

File: server/routes/bookedtoday/arrivals/searchforexisting/multipleExistingRecordsFoundController.test.ts [hunk 6]
```diff
@@ -41,7 +43,7 @@ beforeEach(() => {
   expectedArrivalsService.getMatchingRecords.mockResolvedValue(potentialMatches)
   expectedArrivalsService.getPrisonerDetails.mockResolvedValue(potentialMatches[0])
   flashProvider.mockReturnValue([])
-  signedCookiesProvider.mockReturnValue({ 'search-details': searchDetails })
+  stubCookie(State.searchDetails, searchDetails)
```
Comment (andrewrlee) [inline server/routes/bookedtoday/arrivals/searchforexisting/multipleExistingRecordsFoundController.test.ts:46]: this now has type safety - searchDetails needs to have same type as `SearchDetails`

File: server/routes/bookedtoday/arrivals/searchforexisting/multipleExistingRecordsFoundController.test.ts [hunk 7]
```diff
@@ -50,14 +52,6 @@ afterEach(() => {
 
 describe('possible records found', () => {
   describe('view', () => {
-    it('should get search details from state', () => {
-      return request(app)
-        .get('/prisoners/12345-67890/search-for-existing-record/possible-records-found')
-        .expect(() => {
-          expect(signedCookiesProvider).toHaveBeenCalledTimes(1)
-        })
-    })
```
Comment (andrewrlee) [inline server/routes/bookedtoday/arrivals/searchforexisting/multipleExistingRecordsFoundController.test.ts:59]: don't need to test this implementation detail

File: server/routes/bookedtoday/arrivals/searchforexisting/multipleExistingRecordsFoundController.test.ts [hunk 8]
```diff
@@ -101,10 +95,18 @@ describe('possible records found', () => {
...
-          expect(signedCookiesProvider).toHaveBeenCalledTimes(1)
+          expectSettingCookie(res, State.newArrival).toStrictEqual({
+            dateOfBirth: '1973-01-08',
+            expected: 'true',
+            firstName: 'James',
+            lastName: 'Smyth',
+            pncNumber: '11/5678',
+            prisonNumber: 'A1234BC',
+            sex: 'MALE',
+          })
```
Comment (andrewrlee) [inline server/routes/bookedtoday/arrivals/searchforexisting/multipleExistingRecordsFoundController.test.ts:109]: do need to test this side effect

File: server/routes/bookedtoday/arrivals/searchforexisting/search/changeDateOfBirthController.test.ts [hunk 9]
```diff
@@ -49,8 +50,6 @@ describe('GET /search-for-existing-record/change-date-of-birth', () => {
   })
 
   it('redirects when no cookie present', () => {
-    signedCookiesProvider.mockReturnValue({})
```
Comment (andrewrlee) [inline server/routes/bookedtoday/arrivals/searchforexisting/search/changeDateOfBirthController.test.ts:52]: if no cookie then don't need to say so

File: server/routes/bookedtoday/arrivals/searchforexisting/singleExistingRecordFoundController.test.ts [hunk 10]
```diff
@@ -1,33 +1,41 @@
...
+      State.searchDetails,
+      {
+        firstName: 'James',
+        lastName: 'Smyth',
+        dateOfBirth: '1973-01-08',
+        prisonNumber: undefined,
+        pncNumber: '01/98644M',
+      },
+    ],
+  ])
```
Comment (andrewrlee) [inline server/routes/bookedtoday/arrivals/searchforexisting/singleExistingRecordFoundController.test.ts:38]: example of stubbing multiple cookies

File: server/routes/bookedtoday/arrivals/state.ts [hunk 11]
```diff
@@ -8,16 +8,29 @@ export type NewArrival = {
...
+    lastName: value.lastName,
+    dateOfBirth: value.dateOfBirth,
+    expected: value.expected.toString(),
+    ...(value.prisonNumber !== undefined && { prisonNumber: value.prisonNumber }),
+    ...(value.pncNumber !== undefined && { pncNumber: value.pncNumber }),
+    ...(value.sex !== undefined && { sex: value.sex }),
+    ...(value.code !== undefined && { code: value.code }),
+    ...(value.imprisonmentStatus !== undefined && { imprisonmentStatus: value.imprisonmentStatus }),
+    ...(value.movementReasonCode !== undefined && { movementReasonCode: value.movementReasonCode }),
+  }),
```
Comment (andrewrlee) [inline server/routes/bookedtoday/arrivals/state.ts:30]: now being more careful around writing cookie to state - previously we would push anything through. 
This meant we were incorrectly setting cro number in the cookie. Fixed this.

File: server/routes/bookedtoday/arrivals/state.ts [hunk 12]
```diff
@@ -66,6 +80,6 @@ const SearchDetailsCodec: Codec<SearchDetails> = {
 }
 
 export const State = {
-  searchDetails: stateOperations('search-details', SearchDetailsCodec),
-  newArrival: stateOperations('new-arrival', NewArrivalCodec),
+  searchDetails: new StateOperations('search-details', SearchDetailsCodec),
+  newArrival: new StateOperations('new-arrival', NewArrivalCodec),
```
Comment (andrewrlee) [inline server/routes/bookedtoday/arrivals/state.ts:84]: Now a class so can more easily access type info compared to an anonymous function

File: server/routes/bookedtoday/unexpectedArrivals/multipleExistingRecordsFoundController.ts [hunk 13]
```diff
@@ -35,7 +35,7 @@ export default class MultipleExistingRecordsFoundController {
       const { prisonNumber } = req.body
       const selectedRecord = await this.expectedArrivalsService.getPrisonerDetails(prisonNumber)
 
-      State.newArrival.set(res, selectedRecord)
+      State.newArrival.set(res, { ...selectedRecord, expected: false })
```
Comment (andrewrlee) [inline server/routes/bookedtoday/unexpectedArrivals/multipleExistingRecordsFoundController.ts:38]: not expected!

**Review by louisenorris — APPROVED** [review]
LGTM

**Pull request merged**